### DISCUSSION TITLE
Daily and Weekly Notes - any day from calendar

### DISCUSSION THREAD

**Discussion by Callumk7** [SilentVoid13/Templater#568] [Templates Showcase]
**Problem**: Using tp.date.now wasn't compatible with making future (or past) daily or weekly notes. 

**Solution**: moment.js

**Daily**
```javascript
//breadcrumbs frontmatter
Next: [[<% moment(tp.file.title, "DD-MM-YYYY").add(1, "days").format("DD-MM-YYYY") %>]]
Prev: [[<% moment(tp.file.title, "DD-MM-YYYY").subtract(1, "days").format("DD-MM-YYYY") %>]]

//title
# <% moment(tp.file.title, "DD-MM-YYYY").format("Do MMMM YYYY") %>
```
**Notes**
I use format DD-MM-YYYY, which has to be specified in moment as above. 
You must set daily notes to be created using the same format as specified in the template.

**Weekly**
```javascript
//breadcrumbs frontmatter
Next: [[Week <% moment(tp.file.title, "ww YYYY").add(1, "weeks").format("ww YYYY") %>]]  
Prev: [[Week <% moment(tp.file.title, "ww YYYY").subtract(1, "weeks").format("ww YYYY") %>]]

//title
# Week <% moment(tp.file.title, "ww YYYY").format("ww YYYY") %>

//rename title script
<% tp.file.rename("Week" + tp.file.title) %>
```

**Notes**
I name my weekly notes "Week 01 2020" for example, which requires an additional step of wrangling. I set the calendar plugin to create notes in the format "ww YYYY". The included rename title script above sits at the bottom of my weekly template, and adds the "Week" string to the file title.

**Comment by ChauThan** [comment]
I got this from digital.graden template. Hope it is useful.
```
<% tp.user.generate_periodic_link_markdown(tp, "D") %>
```

Result
<< [Yesterday](app://obsidian.md/2022-03-20) | [Tomorrow](app://obsidian.md/2022-03-22) >>

**Comment by AtlcatlQ** [comment]
I was about to post it 😂, I just realize of that today.

**Comment by NeodymiumPhish** [comment]
```
Templater Error: Template parsing error, aborting. 
 tp.user.generate_periodic_link_markdown is not a function
```

That's odd. Templater says that's not a valid function. Can someone provide guidance?

**Reply by ChauThan** [reply]
I guess you are missing Periodic notes plugin

**Reply by NeodymiumPhish** [reply]
I do, actually. :/

```
---
tags: daily-note
full-date: <% tp.date.now() %>
week: <% tp.date.now("YYYY-[W]ww", 0, tp.file.title, "YYYY-MM-DD") %>
month: <% tp.date.now("YYYY-[M]MM", 0, tp.file.title, "YYYY-MM-DD") %>
year: <% tp.date.now("YYYY", 0, tp.file.title, "YYYY-MM-DD") %>
---


<% tp.user.generate_periodic_link_markdown(tp, "D") %>


# Meeting Notes


# Overview
```

^ That's the template I'm using for daily notes (it's a work in progress).

And this is what it shows when I create the template:

<img width="480" alt="Screen Shot 2022-03-26 at 9 28 37 PM" src="https://user-images.githubusercontent.com/7989828/160262824-d49e2865-d878-4af9-871f-66f91eee1291.png">

**Reply by ChauThan** [reply]
Sorry, 
I thought it is a script of `Periodic notes plugin`.
But it is a script of `digital.garden` author.
You can read in [generate_periodic_link_markdown.js](https://github.com/tanepiper/obsidian-garden/blob/main/_scripts/generate_periodic_link_markdown.js)

**Reply by ChauThan** [reply]
You have to configure `_scripts` folder in `Templater`'s settings.
![image](https://user-images.githubusercontent.com/5886174/160373010-921ee82e-6958-4a9f-b522-25c271c0aa38.png)

**Comment by JeromeZephyr** [comment]
I do not understand why you say you need moment.js.

I did it this way.

# Navigation template
```
< [[<% tp.date.now("YYYY-MM-DD",-1,tp.file.title,"YYYY-MM-DD") %>]] | [[<% tp.date.now("GGGG-[W]WW",7,tp.file.title,"YYYY-MM-DD") %>]] | [[<% tp.date.now("YYYY-MM-DD",1,tp.file.title,"YYYY-MM-DD") %>]] >
```

For me, the name of the note is the curent date. You could remove everything after the -1 and +1. 

# Result
```
< [[<% tp.date.now("YYYY-MM-DD",-1) %>]] | [[<% tp.date.now("GGGG-[W]WW",7) %>]] | [[<% tp.date.now("YYYY-MM-DD",1) %>]] >
```

**Reply by Callumk7** [reply]
That only works for today. The problem we are solving is stated in the original message.

**Reply by JeromeZephyr** [reply]
The code after *Navigation template* work with the future if you note is name like **YYYY-MM-DD**

**Reply by Volker-brdb** [reply]
with moment it works in the past and the future.
e.g. 
[[<% moment(tp.file.title).add(-1, 'day').format("YYYY-MM-DD")  %>]] | <-- today--> |  [[<% moment(tp.file.title).add(1, 'day').format("YYYY-MM-DD)  %>]]

you can create a note for eg 21st July 2021 and everything is fine and even if creating one in the future for e.g. 1st Sep 2023

**Reply by JeromeZephyr** [reply]
Indeed, you can achieve this using moment. 
I just want to point that you can acheve the same result using parameters of the Date Module of Templater. 
Reference: [tp.date - Templater](https://silentvoid13.github.io/Templater/internal-functions/internal-modules/date-module.html#tpdatenowformat-string--yyyy-mm-dd-offset-numberstring-reference-string-reference_format-string)

The *now* in `tp.date.now` might be misleading because it can be used to work with any date by using the third parameter. 

The code below gives the same result as your example. 
```
[[<% tp.date.now("YYYY-MM-DD",-1,tp.file.title,"YYYY-MM-DD") %>]] | <-- today--> | [[<% tp.date.now("YYYY-MM-DD",1,tp.file.title,"YYYY-MM-DD") %>]]
```

**Comment by HyperCrusher** [comment]
The periodic notes plugin also fixes this using {{date:}} and you can plug that into tp.now
`tp.date.now("YYYY/MMMM/Do", -1," {{date:MMMM/D/YYYY}}")`
will always output Year/Month/Day of the note before this one, no matter what day you select on the calendar plugin because it uses the periodic notes date replace

**Reply by Kmfernan5** [reply]
Can you use daily notes periodic notes templater random note all together

**Comment by slambert** [comment]
I'm trying to get this working for a monthly template, but encountering invalid dates.

Here's how it works (and doesn't work)

This is an array to pick the month you're creating:
```
---
<%*
const month = [
{ label: "Jan", value: "01" },
{ label: "Feb", value: "02" },
{ label: "Mar", value: "03" },
{ label: "Apr", value: "04" },
{ label: "May", value: "05" },
{ label: "Jun", value: "06" },
{ label: "Jul", value: "07" },
{ label: "Aug", value: "08" },
{ label: "Sep", value: "09" },
{ label: "Oct", value: "10" },
{ label: "Nov", value: "11" },
{ label: "Dec", value: "12" }
];
```

This renames the file using the array above and tp.system.suggester
```
const renameMonthYear = await tp.system.suggester(option => option.label, month);
tR += await tp.file.rename("2023-" + renameMonthYear.value);
-%>
```

All that works.

Then I try to do something similar with the frontmatter. But the MMM-YYYY tag kicks back an invalid date:

```
creation date: <% tp.file.creation_date() %>
tags: MonthlyNote <%* tR += moment(tp.file.title, "YYYY MM").format("MMM-YYYY") %>  y<% tp.file.creation_date("YYYY") %>
---
```

And just trying to output a reformatted date based on the year month or 2023-10 format, I'm getting invalid dates.

```
<%* tR += moment(tp.file.title, "YYYY MM").format("MMMM gggg") %>
```

I'm sure I'm missing something small...

**Reply by Zachatoo** [reply]
Would you happen to be renaming the file and accessing `tp.file.title` later in the note?

At the moment, renaming a file does not update `tp.file.title` during that template execution. So it's likely looking at what the note used to be named before the rename.

To get around this, you can put what you want to rename your note to in a variable in an [execution command](https://silentvoid13.github.io/Templater/commands/execution-command.html) and reference that instead of `tp.file.title` later in the template.

```js
<%*
// ...
const renameMonthYear = await tp.system.suggester(option => option.label, month);
const title = "2023-" + renameMonthYear.value; // Use title instead of tp.file.title
await tp.file.rename(title);
-%>
---
creation date: <% tp.file.creation_date() %>
tags: MonthlyNote <% moment(title, "YYYY MM").format("MMM-YYYY") %>  y<% tp.file.creation_date("YYYY") %>
---
